File: Sorting/insertion_sort.py
def insertion_sort(arr):
    """
    Perform Insertion Sort on a list with detailed output.
    
    Parameters:
    arr (list): The list of elements to sort.

    Returns:
    None: The list is sorted in place.
    """
    print(f"Original array: {arr}")  # Print the original array

    for i in range(1, len(arr)):
        key = arr[i]
        j = i - 1
        
        # Show the key being considered
        print(f"\nKey being inserted: {key}")
        
        # Shift elements of the sorted portion to the right
        while j >= 0 and arr[j] > key:
            arr[j + 1] = arr[j]
            j -= 1
            print(f"  After shifting: {arr}")  # Show the array after each shift

        # Insert the key at the correct position
        arr[j + 1] = key
        print(f"  Inserted {key} at position {j + 1}: {arr}")  # Show the array after insertion

    print(f"\nSorted array: {arr}")  # Print the final sorted array


# Example usage
if __name__ == "__main__":
    arr = [12, 11, 13, 5, 6]  # Test array
    insertion_sort(arr)  # Perform the sorting

File: Sorting/insertion_sort.cpp
#include <iostream>
using namespace std;

// Function to perform Insertion Sort
void insertionSort(int arr[], int n) {
    // Traverse the array starting from the second element
    for (int i = 1; i < n; i++) {
        int key = arr[i];  // The current element to be placed in the sorted portion
        int j = i - 1;     // Start comparing from the last element of the sorted portion

        // Shift elements of the sorted portion to the right
        // until the correct position for the key is found
        while (j >= 0 && arr[j] > key) {
            arr[j + 1] = arr[j];  // Move the larger element one position to the right
            j--;                  // Move to the previous element in the sorted portion
        }

        // Place the key in its correct position
        arr[j + 1] = key;
    }
}

// Function to print the array
void printArray(int arr[], int n) {
    for (int i = 0; i < n; i++) {
        cout << arr[i] << " ";
    }
    cout << endl;
}

// Main function to test the Insertion Sort
int main() {
    int arr[] = {12, 11, 13, 5, 6};  // Example array to be sorted
    int n = sizeof(arr) / sizeof(arr[0]);  // Calculate the size of the array

    cout << "Original array: ";
    printArray(arr, n);

    insertionSort(arr, n);  // Perform the sorting

    cout << "Sorted array: ";
    printArray(arr, n);

    return 0;
}
